docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc

```diff
@@ -262,6 +260,16 @@ Users of the `pbkdf2` provider might need to explicitly reduce the
 number of hash iterations to regain acceptable performance.
 This can be done by configuring the hash iterations explicitly in the password policy of the realm.
 
+== Expected increased overall CPU usage and temporary increased database activity
+
+The Concepts for sizing CPU and memory resources in the Keycloak High Availability guide have been updated to reflect the new hashing defaults.
+While the CPU usage per password-based login in our tests increased by 33% (which includes both the changed password hashing and unchanged TLS connection handling), the overall CPU increase should be around 10% to 15%.
+This is due to the averaging effect of Keycloak's other activities like refreshing access tokens and client credential grants, still this depends on the unique workload of an installation.
+
+After the upgrade, during a password-based login, the user's passwords will be re-hashed with the new hash algorithm and hash iterations as a one-off activity and updated in the database.
+As this clears the user from Keycloak's internal cache, you'll also see an increased read activity on the database level.
+This increased database activity will decrease over time as more and more user's passwords have been re-hashed.
+
```

Comment:
@kami619  / @thomasdarimont  / @pedroigor - We've run tests with the new hashing, and these are the results. I've added them to the migration guide. 

Let me know what you think about this. This could also be part of a blog post around hashing if that is still planned.

I've be happy to hear your comments if this can be merged as is, or if it requires additional changes.

Reply:
Thanks for adding the note! This will surely help users to better understand the impact. However, I wonder if ALL customers are able to adapt this so easily... customers who run Keycloak in a managed environment that can be easily scaled might not have a problem with this, but customers who run Keycloak with dedicated hardware might need some time to adapt.

How about introducing a server-side configuration flag like `spi.credential-keycloak-password-enforce-password-policy-on-validation` (default=true) for the `org.keycloak.credential.PasswordCredentialProvider` that can be used to disable the rehashing during password based login? The new hashing rules would then then still be applied on password update or creation.
This would give customers at least a way to quickly mitigate load if they detect that they need more CPU capacity.

The flag could be checked in: `org.keycloak.credential.PasswordCredentialProvider#isValid(..)`.